### PR TITLE
Add more TextStyle support to Paragraph in CanvasKit mode

### PR DESCRIPTION
**Pull request by hterkelsen**
## Description

* Adds support for StrutStyle in ParagraphStyle.

Adds support for the following in TextStyle:
* decoration color
* decoration style
* text baseline
* letter spacing
* word spacing
* height
* locale
* shadows
* font features

* Adds support for addPlaceholder in ParagraphBuilder

## Related Issues

Fixes https://github.com/flutter/flutter/issues/63690
Fixes https://github.com/flutter/flutter/issues/63616
Fixes https://github.com/flutter/flutter/issues/63144
Fixes https://github.com/flutter/flutter/issues/61277
Fixes https://github.com/flutter/flutter/issues/61240
Fixes https://github.com/flutter/flutter/issues/58491
Fixes https://github.com/flutter/flutter/issues/67993

## Tests

Updated `canvaskit_api_test.dart` to use the new SkRects.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvaskit/text.dart [hunk 1]
```diff
@@ -423,7 +627,44 @@ class CkParagraphBuilder implements ui.ParagraphBuilder {
     double? baselineOffset,
     ui.TextBaseline? baseline,
   }) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/text.dart:629]: You can remove the TODO above the method

File: lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart [hunk 2]
```diff
@@ -79,6 +79,10 @@ class CanvasKit {
   external int get LineThroughDecoration;
   // End of text decoration enum.
 
+  external SkTextDecorationStyleEnum get DecorationStyle;
+  external SkTextBaselineEnum get TextBaseline;
+  external SkPlaceholderAlignmentEnum get PlaceholderAlignment;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart:84]: Let's add tests for all of the new bindings in `canvaskit_api_test.dart`.

File: lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart [hunk 3]
```diff
@@ -1425,8 +1376,10 @@ class SkCanvas {
   external int save();
   external int getSaveCount();
   external void saveLayer(
-    SkRect bounds,
-    SkPaint paint,
+    SkPaint? paint,
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart:1379]: Wow, was this a bug before or did the parameter order flip in the new version?

Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart:1379]: The order flipped in the new version

File: lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart [hunk 4]
```diff
@@ -1547,14 +1564,85 @@ class SkTextStyleProperties {
   external double? get decorationThickness;
   external set decorationThickness(double? value);
 
+  external Float32List? get decorationColor;
+  external set decorationColor(Float32List? value);
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart:1568]: We could remove all the getters. Are they used anywhere?

File: lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart (hunk 2 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart:84]: Done.

File: lib/web_ui/lib/src/engine/canvaskit/text.dart [hunk 5]
```diff
@@ -187,21 +296,103 @@ class CkTextStyle implements ui.TextStyle {
...
-    //   - decorationStyle
-    //   - textBaseline
-    //   - letterSpacing
-    //   - wordSpacing
-    //   - height
-    //   - locale
-    //   - shadows
-    //   - fontFeatures
+    if (shadows != null) {
+      var ckShadows = <SkTextShadow>[];
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/text.dart]: Done.

File: lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart [hunk 6]
```diff
@@ -1015,12 +1019,12 @@ class SkPath {
   external SkPath([SkPath? other]);
   external void setFillType(SkFillType fillType);
   external void addArc(
-    SkRect oval,
+    List<double> oval,
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart]: Done.

**Comment by hterkelsen** [comment]
Thanks, Yegor. PTAL. I suggest we just remove the getters from the `Sk` types and basically treat them as black boxes unless they're actually used somewhere.

File: lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart:1578]: If not let's remove them.

**Review by yjbanov — APPROVED** [review]
Leaving the decision to remove the unnecessary getters to you.

lgtm

**Comment by hterkelsen** [comment]
Thanks, I got rid of the getters since they're unused and could potentially cause problems down the line.

**Pull request merged**